### PR TITLE
.murdock: Add modules/pkgs diff of make/kconfig

### PR DESCRIPTION
**Pull request by MrKevinWeiss**
### Contribution description

When there is a difference between the make dependency resolution and kconfig resolution (ie. `TEST_KCONFIG=1`) this gets murdock to show the diff of the modules and packages.
This gives clearer feedback than the binary mismatch error message.
Note that this comparison only occurs when there is a mismatch and shouldn't affect the a passing PR.

### Testing procedure

I will add a failure commit so it should be viewable from murdock.
One can also check with
```
/bin/bash -c "source .murdock; JOBS=4 compile ${TEST_DIR} ${BOARD}:gnu"
```

or with the `REMOVEME!!!` commit

```
/bin/bash -c "source .murdock; JOBS=4 compile examples/hello-world native:gnu"
```

### CODE REVIEW

**Comment by MrKevinWeiss** [comment]
Should I just save the output in `/tmp/*` instead of using `<(...)`?  Any opinions, again, locally works but on murdock it shows `./.murdock:/Syntax/error:/"("/unexpected`

**Comment by leandrolanzieri** [comment]
murdock still seems not to like this

File: .murdock [hunk 1]
```diff
@@ -107,6 +107,17 @@ get_supported_kconfig_board_app() {
...
 
+kconfig_module_packages_diff() {
+    local board=$1
+    local appdir=$2
+    make clean --no-print-directory -C ${appdir} BOARD=${board}
+    local make_modules=$(make --no-print-directory info-modules info-packages -C ${appdir} BOARD=${board})
+    make clean --no-print-directory -C ${appdir} BOARD=${board}
+    local kconfig_modules=$(TEST_KCONFIG=1 make --no-print-directory info-modules info-packages -C ${appdir} BOARD=${board})
+
+    diff <(echo "${make_modules}") <(echo "${kconfig_modules}")
```
Comment (MrKevinWeiss) [inline .murdock:118]: I guess I cannot use this with the .murdock scripts...  I will just write a tmp_file or something and do a diff on that...

**Comment by MrKevinWeiss** [comment]
ping @kaspar030 @kfessel  @leandrolanzieri 

This contains the fast failing when kconfig modules/packages are not matching (ie skip compile) as well as.

May I squash and remove the demo commit?

**Comment by MrKevinWeiss** [comment]
This is getting pretty heavy now...  I added some app/board namespacing for the file creation to try to prevent any possible race conditions.  The best would be if I could get the `diff <(make ...) <(TEST_KCONFIG=1 make)` stuff working but murdock seems to have some issues with that.

Once we approve of everything I will run some tests (at night when murdock is not busy) to see how much overhead it adds to a successful run vs not).

It would be nice if @kaspar030 took a quick look.

**Comment by MrKevinWeiss** [comment]
hmmm it appears the packages get cloned causing messages that make the diff fail.

**Comment by kfessel** [comment]
would it somehow be possible to trigger this with make or a shell script outside of . murdock?

**Comment by MrKevinWeiss** [comment]
Currently I am  not liking the slowdown:

<details>

```
weiss@mobiweiss:~/wd/RIOT$ hyperfine -m 30 '/bin/bash -c "source .murdock; compile examples/hello-world native:gnu"'
Benchmark #1: /bin/bash -c "source .murdock; compile examples/hello-world native:gnu"
  Time (mean ± σ):      2.265 s ±  0.337 s    [User: 4.342 s, System: 1.131 s]
  Range (min … max):    1.953 s …  3.257 s    30 runs
 
  Warning: Statistical outliers were detected. Consider re-running this benchmark on a quiet PC without any interferences from other programs. It might help to use the '--warmup' or '--prepare' options.
 
weiss@mobiweiss:~/wd/RIOT$ git checkout master
Switched to branch 'master'
Your branch is ahead of 'origin/master' by 1033 commits.
  (use "git push" to publish your local commits)
weiss@mobiweiss:~/wd/RIOT$ hyperfine -m 30 '/bin/bash -c "source .murdock; compile examples/hello-world native:gnu"'
Benchmark #1: /bin/bash -c "source .murdock; compile examples/hello-world native:gnu"
  Time (mean ± σ):      1.550 s ±  0.071 s    [User: 3.520 s, System: 0.900 s]
  Range (min … max):    1.473 s …  1.711 s    30 runs

```
</details>

2.265 with pr, 1.550 without

I think we need to find a way around filesystem writing...  Maybe just compare the modules in make and don't use diff or something.

**Comment by MrKevinWeiss** [comment]
> would it somehow be possible to trigger this with make or a shell script outside of . murdock?

Currently no, the whole thing would involve calling make several times.  I would be OK adding it to dist/tools/ci, maybe. We should be careful with the speed though, since this only helps if something will fail, the pass case should not take too much longer.

First thing would be to speed it up.

**Comment by MrKevinWeiss** [comment]
Hmmm I think that we cannot actually do the failfast in the CI, simply calling the `info-modules info-packages` twice with cleaning adds an extra ~30% to the job time (for hello world, which would be less for bigger binaries but still).

How about printing the diffs on failure and adding a script that can be run on your local machine to do a check (I don't know how long that would take though)?

**Comment by MrKevinWeiss** [comment]
Even better lets keep this PR simple and just do what I wanted to in the beginning which is print on failure.

I think before I make a script to do a kconfig test deps I would first like to get #16496 in as it would play a big part I guess... Or maybe move the list that is in the .murdock file out to use it...

**Comment by leandrolanzieri** [comment]
The output shows the diff as expected, please remove 258efb7 @MrKevinWeiss

**Review by leandrolanzieri — APPROVED** [review]
ACK

**Comment by MrKevinWeiss** [comment]
Many thanks!

**Pull request merged**